### PR TITLE
Add shellcheck step to PR workflow.

### PR DESCRIPTION
**Pull request by jmczerk**
Enables PR check to lint shell scripts with `shellcheck`.  Script changes are needed to get a clean linter run.

These changes have been tested with AI Notebook, Dataproc Cluster, and custom apps.

### CODE REVIEW

**Comment by jmczerk** [comment]
Ignoring [SC1090](https://www.shellcheck.net/wiki/SC1090), [SC1091])(https://www.shellcheck.net/wiki/SC1091) out of the gate, as `shellcheck` seems to have limitations around dynamic file resolution that lead to lots of false positives.

Distribution of existing errors:

40 x [SC2155 - Declare and assign separately to avoid masking return values.](https://www.shellcheck.net/wiki/SC2155)
12 x [SC2086 - Double quote to prevent globbing and word splitting.](https://www.shellcheck.net/wiki/SC2086)
5 x [SC2034 - foo appears unused. Verify it or export it.](https://www.shellcheck.net/wiki/SC2034)
1 x [SC2016 - Expressions don't expand in single quotes, use double quotes for that.](https://www.shellcheck.net/wiki/SC2016)
1 x [SC2046 - Quote this to prevent word splitting.](https://www.shellcheck.net/wiki/SC2046)
1 x [SC2145 - Argument mixes string and array. Use * or separate argument.](https://www.shellcheck.net/wiki/SC2145)
1 x [SC2164 - Use `cd ... || exit` in case cd fails.](https://www.shellcheck.net/wiki/SC2164)
1 x [SC2207 - Prefer `mapfile` or `read -a` to split command output (or quote to avoid splitting).](https://www.shellcheck.net/wiki/SC2207)

**Comment by mbookman** [comment]
Thanks @jmczerk ! 

I am a big fan of this, at least conceptually. I haven't used `shellcheck`, so I don't know if we'd find ourselves battling various cases. That's usually a low-cost battle for the upside, though.

I've hesitated to enforce [SC2155](https://www.shellcheck.net/wiki/SC2155) as we declare/execute usually on well-exercised code paths, but enforcing it would avoid a category of painful bugs creeping in.

+1!

**Review by mbookman — CHANGES_REQUESTED** [review]

File: startupscript/git-setup.sh [hunk 1]
```diff
@@ -12,13 +12,15 @@
...
-# - USER_SSH_DIR: path to ssh directory (~/.ssh)
+# - WORK_DIRECTORY: home directory for the user that the script is running on behalf of
 # - WORKBENCH_GIT_REPOS_DIR: path to the git repo directory (~/repos)
 # - RUN_AS_LOGIN_USER: run command as app user
 
 emit "Setting up git integration..."
 
 # Create the user SSH directory
+readonly USER_SSH_DIR="${WORK_DIRECTORY}/.ssh"
+readonly WORKBENCH_GIT_REPOS_DIR="${WORK_DIRECTORY}/repos"
```
[the remark this answers is not in the document]
Comment (jmczerk) [reply startupscript/git-setup.sh:23]: I think there may be value in this; could we maybe capture a Jira ticket to do this in a follow-on?  

I've tried to minimize structural/logic changes in this PR given the large number of changes, and the test cycles across Vertex AI + Dataproc + Mikey custom apps has stretched this beyond the time I budgeted for it.  Are you OK with deferring?

[the remark this answers is not in the document]
Comment (jmczerk) [reply startupscript/git-setup.sh:23]: @mbookman buming this question

**Review by mbookman — APPROVED** [review]
LGTM

File: .github/workflows/test-pr.yaml [hunk 2]
```diff
@@ -3,6 +3,16 @@ on:
...
 jobs:
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ludeeus/action-shellcheck@2.0.0
+        with:
+          check_together: 'yes'
+        env:
+          SHELLCHECK_OPTS: -e SC1090 -e SC1091
```
Comment (rogerwangcs) [inline .github/workflows/test-pr.yaml:14]: Could we add a comment that describes these two lint rules?

**Review by rogerwangcs — APPROVED** [review]
lgtm! Are there any local dev environment setup to properly enable shellcheck? Is it a command we run? Is it integrated with vscode/intellij? If so, it may be worth adding to a `CONTRIBUTINGmd`

**Comment by jmczerk** [comment]
> lgtm! Are there any local dev environment setup to properly enable shellcheck? Is it a command we run? Is it integrated with vscode/intellij? If so, it may be worth adding to a `CONTRIBUTING.md`

There's a new section in `startupscripts/README.md` for this, as that seemed to be where all the startup script docs lived.

**Comment by rogerwangcs** [comment]
> There's a new section in `startupscripts/README.md` for this, as that seemed to be where all the startup script docs lived.

Missed this, thanks.

**Pull request merged**